packages/block-library/src/table-of-contents/utils.js

```diff
@@ -0,0 +1,135 @@
+/**
+ * WordPress dependencies
+ */
+import { select } from '@wordpress/data';
+import { cleanForSlug } from '@wordpress/editor';
+
+export function linearToNestedList( array, index = 0 ) {
+	const returnValue = [];
+
+	array.forEach( function( heading, key ) {
+		if ( typeof heading.content === 'undefined' ) {
+			return;
+		}
+
+		// Make sure we are only working with the same level as the first iteration in our set.
+		if ( heading.level === array[ 0 ].level ) {
+			// Check that the next iteration will return a value.
+			// If it does and the next level is greater than the current level,
+			// the next iteration becomes a child of the current interation.
+			if (
+				typeof array[ key + 1 ] !== 'undefined' &&
+				array[ key + 1 ].level > heading.level
+			) {
+				// We need to calculate the last index before the next iteration that has the same level (siblings).
+				// We then use this last index to slice the array for use in recursion.
+				// This prevents duplicate nodes.
+				let endOfSlice = array.length;
+				for ( let i = key + 1; i < array.length; i++ ) {
+					if ( array[ i ].level === heading.level ) {
+						endOfSlice = i;
+						break;
+					}
+				}
+
+				// We found a child node: Push a new node onto the return array with children.
+				returnValue.push( {
+					block: heading,
+					index: index + key,
+					children: linearToNestedList(
+						array.slice( key + 1, endOfSlice ),
+						index + key + 1
+					),
+				} );
+			} else {
+				// No child node: Push a new node onto the return array.
+				returnValue.push( {
+					block: heading,
+					index: index + key,
+					children: null,
+				} );
+			}
+		}
+	} );
+
+	return returnValue;
+}
+
+export function getPageHeadings() {
+	return convertHeadingBlocksToAttributes( getHeadingBlocks() );
+}
+
+export function getHeadingBlocks() {
+	const editor = select( 'core/block-editor' );
+	return editor
+		.getBlocks()
+		.filter( ( block ) => block.name === 'core/heading' );
+}
+
+export function convertHeadingBlocksToAttributes( headingBlocks ) {
+	return headingBlocks.map( function( heading ) {
+		const level = heading.attributes.level.toString();
+
+		const headingContent = heading.attributes.content || '';
+		const anchorContent = heading.attributes.anchor || '';
+
+		// strip html from heading and attribute content
+		const contentDiv = document.createElement( 'div' );
+
+		contentDiv.innerHTML = headingContent;
+		const content = contentDiv.textContent || contentDiv.innerText || '';
+
+		contentDiv.innerHTML = anchorContent;
+		let anchor = contentDiv.textContent || contentDiv.innerText || '';
+
+		if ( anchor !== '' && anchor.indexOf( '#' ) === -1 ) {
```

Comment:
This can be made more clear using [`String.prototype.includes`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/includes):
```suggestion
		if ( anchor !== '' && ! anchor.includes( '#' ) ) {
```